### ISSUE TITLE
Model explanations for regression problems

### ISSUE THREAD

**Issue by arnaudvl** [SeldonIO/alibi#19]
Make explainer modules work for regression problems using e.g. binning for the prediction.

**Comment by frederikfaye** [comment]
I'd love this feature! 

Is anyone working on this? What would be the scope of implementing this?

**Comment by jklaise** [comment]
Thanks for your interest @frederikfaye ! Support for regression problems is quite an open-ended area and there are many ways to do this, so the scope can be quite broad (e.g. even doing it the obvious way of binning the continuous response into discrete bins can be done in various ways depending on the use case). Do you have any particular use case that you have in mind?

**Comment by frederikfaye** [comment]
When you say
> binning the continuous response into discrete bin

do you mean such that the problem effectively becomes a multiclass classification problem?

I wanted to explain a trained CNN, and I was originally looking at something like a modified GradCAM (e.g. [here](https://jacobgil.github.io/deeplearning/vehicle-steering-angle-visualizations) (see the "Third attempt" section) or [here](https://github.com/raghakot/keras-vis/tree/master/applications/self_driving)), but the more I look at it, I don't think it would be that informative for my particular problem, as I don't have a naturally defined "center" of my regression target (like angle=0 in the linked-to steering angle problems). I guess SHAP would maybe be a solution, but if you have any other suggestions, please do let me know!

However, I now realize that the thing I would be most interested in would actually be a confidence method for regression models (so sorry that I hijacked this thread, which is about model explanations!). 

As far as I can see, trust scores only work for classification tasks. I've instead been looking at the following two papers, but their implementation is pretty model-dependent: [Bayesian Uncertainty Estimation for Batch Normalized Deep Networks](https://arxiv.org/pdf/1802.06455.pdf) and [A Scalable Laplace Approximation for Neural Networks](https://openreview.net/pdf?id=Skdvd2xAZ). 
Again, do you have any suggestions? Or maybe these are not in the scope of what Alibi seeks to cover?

Best,
Frederik

**Comment by jklaise** [comment]
> do you mean such that the problem effectively becomes a multiclass classification problem?

Yes, if you can find an appropriate way of binning your response variable then you could use all these techniques, e.g. you should be able to use trust scores by binning the response without changing your underlying model.

That being said, confidence methods for regression models are definitely in the roadmap for Alibi, but not the immediate focus. It is a harder problem than classification because if our regressor only returns scalar values there is not much to work with. We are also on the look-out for interesting model-specific (white-box?) methods, in particular the Laplace approx. paper you linked looks very interesting if you have access to the full model.

I think SHAP would be interesting to try, I don't think I've seen an example with regression and image data.

Last note,there is a cool toolbox (https://github.com/albermax/innvestigate) which brings a lot of the methods for inspecting CNNs together (mostly classification though). Some things seem to work better than others, in particular one should be wary of saliency maps for use in explanations (https://arxiv.org/abs/1810.03292) !

**Comment by frederikfaye** [comment]
Thanks for the suggestions, much appreciated! 

I'm curious, what about the Laplace approx. paper caught your interest over the other paper?

**Comment by jklaise** [comment]
Purely because I saw that the other one is BatchNorm specific, so wouldn't generalize to non-image applications! In any case I will have a closer look, also feel free to share any other research you come across as we are on the lookout for good ideas!

**Comment by frederikfaye** [comment]
Right, makes sense - and will do! :)

**Comment by jklaise** [comment]
`ALE`, `IntegratedGradients` and `SHAP` versions all support regression use cases so closing this for now. Will open a new issue if focus on regression use cases specifically becomes a more strategic direction.
